docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+    * syncID - the global syncID for the entire storage managed by sync.
+    * storageVersion - an integer. If this value is greater than a sync client knows about, that client will refuse to sync. While this is rarely changed, it allows us to lock out older sync clients which could be useful.
+    * The list of "declined" engines - that is, engines which the user has explicitly declined from syncing. Note that declined engines are global for the account - declining an engine on any device should decline it on every device. While this is something we'd like to consider changing in the future, for now we are sticking with this global declined state.
+    * Some metadata about every collection (which may be subtly different from an "engine"). Currently this metadata is only {version, syncID}
+
+    Management of meta/global includes all syncIDs - if there's no global syncID, or an engine has no syncID, it should generate one. As we mention below, engines will be responsible for persisting the syncID and version and taking necessary action when it changes. Thus, engines never need to communicate a new syncID back to the global.
+
+* Query info/collections and pass this information to engines.
+
+* Manage interaction with the token server to obtain a token. As part of this, the manager needs to detect when an engine fails to use the token, generally due to a node reassignment mid sync. In this scenario, the manager will abort any pending syncs and reset itself, starting a new sync flow. Note also that this means the containing application will generally not need to see authentication errors from engines, only from the manager.
+
+* Manage the "clients" collection - we probably can't ignore this any longer, especially for bookmarks (as desktop will send a wipe command on bookmark restore). This involves both communicating with the engine regarding commands targetting it, and accepting commands to be send to other devices. Note however that it's not yet clear where these commands will originate from - eg, consider a bookmark restore - which point in this process will queue the wipe commands for other clients?). Exactly how to manage this (eg, when we can assume the engine has processed the command?) are TBD.
+
+* Perform, or coordinate, the actual sync of the engines - from the containing app's POV, there's a single "sync now" entry-point. Exactly how the manager binds to engines is TBD. For example, a possible implementation is that a callback is registered with the sync manager. It's also likely that we will expose a way to sync a specific engine rather than "sync all".
+
+* Manage the fact that there may not be a 1:1 relationship between collections and sync engines. For example, it may be true in the future that a single engine uses multiple collections. A possible example is when we come to syncing "containers" - the history engine may then use 2 collections as part of a sync.
```

Comment:
History and bookmarks are a lot more closely related. Even in my generic record sync plan, different engines in the same backing store can basically be treated as completely distinct.

Reply:
> IIUC we also kind of consider "forms" and "addresses"

I guess you mean "creditcards" and "addresses" - that's an interesting discussion which touches on Thom's "generic store" proposal, but the tl;dr of that is that the implementation of those engines was an exercise in code reuse rather than due to the stores being closely related logically. Stuff like reconciliation is quite separate.

For the purposes of what I wrote in the document above, I'm thinking of a data type which requires 2 collections to have a consistent single local store and is something that would desire cross-collection transaction support on the server.

TBH, I'm still torn as to whether we should support that initially given there's no actual use-case.  On one hand  I suspect containers are going to keep coming up, but on the other hand, it might be enough to have history and containers managed by separate engines so long as we can ensure the order of syncing them.